drivers/SmartThings/harman-luxury/src/disco.lua

```diff
@@ -0,0 +1,130 @@
+local mdns = require "st.mdns"
+local socket = require "cosock.socket"
+local log = require "log"
+
+local disco_helper = require "disco_helper"
+local devices = require "devices"
+
+local SERVICE_TYPE = "_sue-st._tcp"
+local DOMAIN = "local"
+
+local Discovery = {}
+
+local function update_device_discovery_cache(driver, dni, params)
+    log.info("update_device_discovery_cache for device dni: " .. tostring(dni) .. ", " .. tostring(params.ip))
+    local device_info = devices.get_device_info(dni, params)
+    driver.registered_devices[dni] = {
+        ip = params.ip,
+        device_info = device_info
+    }
+end
+
+local function try_add_device(driver, device_dni, device_params)
+    log.trace("try_add_device : dni=" .. tostring(device_dni) .. ", ip=" .. tostring(device_params.ip))
+
+    update_device_discovery_cache(driver, device_dni, device_params)
+
+    local device_info = devices.get_device_info(device_dni, device_params)
+
+    if not device_info then
+        log.error("failed to create device create msg. device_info is nil. dni = " .. device_dni)
+        return nil
+    end
+
+    driver:try_create_device(device_info)
+end
+
+function Discovery.set_device_field(driver, device)
+    log.info("set_device_field : dni = " .. tostring(device.device_network_id))
+    local device_cache_value = driver.registered_devices[device.device_network_id]
+
+    -- persistent fields
+    device:set_field("device_ipv4", device_cache_value.ip, {
```

Comment:
Already mentioned elsewhere but it would be nice if this string (`"device_ipv4`) was captured in a constant somewhere.